exit/test_exit.c

```diff
@@ -0,0 +1,518 @@
+	pid_t pid, gpid;
+	int status;
+
+	gpid = fork();
+	if (gpid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, gpid);
+	}
+	if (gpid == 0) {
+		pid = fork();
+		if (pid < 0) {
+			TEST_ASSERT_LESS_THAN_INT(0, pid);
+		}
+		/* grandchild */
+		if (pid == 0) {
+			/* Sleep for a while to ensure that parent already exited */
+			sleep(4);
+			TEST_ASSERT_NOT_EQUAL_INT(-1, creat(PATH, CREATEMODE));
+			_exit(EXIT_SUCCESS);
+		}
+		/* parent */
+		else {
+			/* parent exits right away */
+			_exit(EXIT_SUCCESS);
+		}
+	}
+	/* grandparent */
+	else {
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+
+		sleep(2);
+		/* Check if file exists, if so child sleep was interrupted */
+		errno = 0;
+		int ret = open(PATH, O_RDWR);
+		if (ret == -1) {
+			TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+		}
+
+		sleep(3);
+		/* At this moment if child wasn't affected file should exist */
+		errno = 0;
+		ret = open(PATH, O_RDWR);
+		TEST_ASSERT_NOT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_NOT_EQUAL_INT(ENOENT, errno);
+
+		close(ret);
+		remove(PATH);
+	}
+}
+
+
+TEST(_exit, new_parent_id)
+{
+	/* Test that child acquire new parent ID */
+	pid_t pid, gpid, ppid;
+	int status;
+
+	gpid = fork();
+	if (gpid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, gpid);
+	}
+	if (gpid == 0) {
+		pid = fork();
+		if (pid < 0) {
+			TEST_ASSERT_LESS_THAN_INT(0, pid);
+		}
+		/* grandchild */
+		if (pid == 0) {
+			ppid = getppid();
+			/* Sleep until parents exits */
+			sleep(2);
+			/* Parent process id should be 1 (init pid) */
+			TEST_ASSERT_NOT_EQUAL_INT(ppid, getppid());
+			_exit(EXIT_SUCCESS);
+		}
+		/* parent */
+		else {
+			/* Sleep until child gets his parent ID */
+			sleep(1);
+			_exit(EXIT_SUCCESS);
+		}
+	}
+	/* grandparent */
+	else {
+		wait(&status);
+		TEST_ASSERT_EQUAL_INT(EXIT_SUCCESS, WEXITSTATUS(status));
+		sleep(3);
+	}
+}
+
+
+TEST(_exit, SIGCHLD_sent)
+{
+	/* Test that SIGCHILD signal is sent after child exits */
+	pid_t pid;
+	int status, ret;
+
+	struct sigaction sa;
+	sa.sa_handler = handler;
+	TEST_ASSERT_EQUAL_INT(0, sigemptyset(&sa.sa_mask));
+	sa.sa_flags = 0;
+	TEST_ASSERT_EQUAL_INT(0, sigaction(SIGCHLD, &sa, NULL));
+
+	errno = 0;
+	ret = open(PATH, O_RDWR);
+	if (ret == -1) {
+		TEST_ASSERT_EQUAL_INT(ENOENT, errno);
+	}
+
+	pid = fork();
+	if (pid < 0) {
+		TEST_ASSERT_LESS_THAN_INT(0, pid);
+	}
+	/* child */
+	if (pid == 0) {
+		/* Exit right away */
+		_exit(EXIT_SUCCESS);
+	}
+	/* parent */
+	else {
+		ret = wait(&status);
+		TEST_ASSERT_EQUAL_INT(pid, ret);
+
+		/* Wait until handler end execution */
+		sleep(1);
+		errno = 0;
+		ret = open(PATH, O_RDWR);
+		TEST_ASSERT_NOT_EQUAL_INT(-1, ret);
+		TEST_ASSERT_NOT_EQUAL_INT(ENOENT, errno);
+
+		TEST_ASSERT_EQUAL_INT(0, close(ret));
+		TEST_ASSERT_EQUAL_INT(0, remove(PATH));
+		sa.sa_handler = SIG_DFL;
+		TEST_ASSERT_EQUAL(0, sigaction(SIGCHLD, &sa, NULL));
+	}
+}
+
+
+TEST_GROUP(exit);
+
+
+TEST_SETUP(exit)
+{
+}
+
+
+TEST_TEAR_DOWN(exit)
+{
+}
+
+
+TEST(exit, stream_flush)
```

Comment:
Please force buffering rules via `setvbuf` to be sure that file writing is unbuffered.

Reply:
I think it is unnecessary to use setvbuf, because fopen with file grants fully buffered stream.

Reply:
Yeah, you`re right. My remark is invalid.